common/achievement/get.ts

```diff
@@ -0,0 +1,99 @@
+import { prisma } from '../prisma';
+import { Achievement_template, User_achievement, achievement_action_type_enum, achievement_type_enum } from '../../graphql/generated';
+import { Step, Achievement, achievement_state } from '../../graphql/types/achievement';
+import { User } from '../user';
+import { renderTemplate } from '../../utils/helpers';
+import { Context } from '../notification/types';
+
+const getUserAchievements = async (user: User): Promise<Achievement[]> => {
+    const userAchievements = await prisma.user_achievement.findMany({
+        where: { userId: user.userID },
+        include: { template: true },
+    });
+    const userAchievementGroups: { [group: string]: User_achievement[] } = {};
+    userAchievements.forEach((ua) => {
+        if (!userAchievementGroups[ua.template.group]) {
+            userAchievementGroups[ua.template.group] = [];
+        }
+        userAchievementGroups[ua.template.group].push(ua);
+    });
+    const achievements: Achievement[] = await createFrontendAchievements(userAchievementGroups, user);
+    return achievements;
+};
+
+const createFrontendAchievements = async (groups: { [group: string]: User_achievement[] }, user: User): Promise<Achievement[]> => {
+    const achievements: Achievement[] = [];
+    for (const group in groups) {
+        const sortedGroupAchievements = groups[group].sort((a, b) => a.groupOrder - b.groupOrder);
+        const groupAchievement: Achievement = await putTogetherAchievement(sortedGroupAchievements, user);
+        achievements.push(groupAchievement);
+    }
+    return achievements;
+};
+
+const putTogetherAchievement = async (userAchievements: User_achievement[], user: User): Promise<Achievement> => {
+    let currentAchievementIndex = userAchievements.findIndex((ua) => !ua.achievedAt);
+    currentAchievementIndex = currentAchievementIndex >= 0 ? currentAchievementIndex : userAchievements.length - 1;
```

Comment:
It does. If we get the index of a userAchievement that has no value in isAchieved we take that index. If we have no index of an achievement since all of them are achieved, we take the index of the last achievementTemplate.